clients/unshaded/src/main/java/tachyon/client/lineage/TachyonLineageFileSystem.java

```diff
@@ -0,0 +1,173 @@
+   * downstream lineages that depend on the given one recursively. Otherwise it does not delete the
+   * lineage exception, if there are other lineages whose input files are the output files of the
+   * specified lineage.
+   *
+   * @param lineageId the id of the lineage
+   * @param cascade whether to delete all the downstream lineages
+   * @return true if the lineage deletion is successful, false otherwise
+   * @throws IOException if the master cannot delete the lineage
+   * @throws LineageDeletionException if the deletion is cascade but the lineage has children
+   * @throws LineageDoesNotExistException if the lineage does not exist
+   */
+  public boolean deleteLineage(long lineageId, boolean cascade)
+      throws IOException, LineageDoesNotExistException, LineageDeletionException {
+    LineageMasterClient masterClient = mContext.acquireMasterClient();
+    try {
+      boolean result = masterClient.deleteLineage(lineageId, cascade);
+      LOG.info(result ? "Succeeded to " : "Failed to" + "delete lineage " + lineageId);
+      return result;
+    } finally {
+      mContext.releaseMasterClient(masterClient);
+    }
+  }
+
+  /**
+   * Lists all the lineages.
+   *
+   * @return the information about lineages
+   * @throws IOException if the master cannot list the lineage info
+   */
+  public List<LineageInfo> listLineages() throws IOException {
+    LineageMasterClient masterClient = mContext.acquireMasterClient();
+
+    try {
+      List<LineageInfo> result = masterClient.listLineages();
+      return result;
+    } finally {
+      mContext.releaseMasterClient(masterClient);
+    }
+  }
+
+  /**
+   * A file is created when its lineage is added. This method reinitializes the created file. But
+   * it's no-op if the file is already completed.
+   *
+   * @return the id of the reinitialized file when the file is lost or not completed, -1 otherwise.
+   * @throws LineageDoesNotExistException if the lineage does not exist.
+   */
+  private long recreate(TachyonURI path, OutStreamOptions options)
+      throws LineageDoesNotExistException {
+    LineageMasterClient masterClient = mContext.acquireMasterClient();
+    try {
+      long fileId =
+          masterClient.recreateFile(path.getPath(), options.getBlockSize(), options.getTTL());
+      return fileId;
+    } catch (IOException e) {
+      throw new RuntimeException("recreation failed", e);
+    } finally {
+      mContext.releaseMasterClient(masterClient);
+    }
+  }
+
+  /**
+   * Gets the output stream for lineage job. If the file already exists on master, returns a dummpy
+   * output stream.
+   */
+  @Override
+  public FileOutStream getOutStream(TachyonURI path, OutStreamOptions options)
+      throws IOException, TachyonException {
+    long fileId;
+    try {
+      fileId = recreate(path, options);
+    } catch (LineageDoesNotExistException e) {
+      // not a lineage file
+      return super.getOutStream(path, options);
```

Comment:
Would it be possible to avoid using exceptions as the control flow? For example `recreate` returns a sentinel value instead of exception?